rs/index/src/ivf.rs

```diff
@@ -1,6 +1,145 @@
+
+use crate::index::Index;
+
+pub struct Ivf {
+    // The whole dataset. TODO(tyb0807): reduce the memory footprint.
+    dataset: Vec<Vec<f32>>,
+    // Number of clusters.
+    num_clusters: usize,
+    // Each cluster is represented by a centroid vector. This is all the centroids in our IVF.
+    centroids: Vec<Vec<f32>>,
+    // Inverted index mapping each cluster to the vectors it contains.
+    //   key: centroid index in `centroids`
+    //   value: vector index in `dataset`
+    inverted_lists: HashMap<usize, Vec<usize>>,
+    // Number of probed centroids.
+    num_probes: usize,
+}
+
+impl Ivf {
+    pub fn new(dataset: Vec<Vec<f32>>, num_clusters: usize, num_probes: usize) -> Self {
+        let config = KMeansConfig::build()
+            .init_done(&|_| debug!("Initialization completed."))
+            .iteration_done(&|s, nr, new_distsum| {
+                debug!(
+                    "Iteration {} - Error: {:.2} -> {:.2} | Improvement: {:.2}",
+                    nr,
+                    s.distsum,
+                    new_distsum,
+                    s.distsum - new_distsum
+                )
+            })
+            .build();
+        let flattened_dataset: Vec<f32> = dataset.iter().flatten().cloned().collect();
+        let dim = dataset[0].len();
+        let kmeans: KMeans<_, 8> = KMeans::new(flattened_dataset, dataset.len(), dim);
+        // TODO(tyb0807): maybe have IvfConfig
+        let result = kmeans.kmeans_minibatch(
+            4, // batch_size
+            num_clusters,
+            1000, // max_iter
+            KMeans::init_random_sample,
+            &config,
+        );
+        debug!("Error: {}", result.distsum);
+        let centroids = result
+            .centroids
+            .chunks(dim)
+            .map(|chunk| chunk.to_vec())
+            .collect();
+
+        let mut inverted_lists = HashMap::new();
+        // Assign vectors to nearest centroids
+        for (i, vector) in dataset.iter().enumerate() {
+            let nearest_centroid = Self::find_nearest_centroid(&vector, &centroids);
+            inverted_lists
+                .entry(nearest_centroid)
+                .or_insert_with(Vec::new)
+                .push(i);
+        }
+        Self {
+            dataset,
+            num_clusters,
+            centroids,
+            inverted_lists,
+            num_probes,
+        }
+    }
+
+    fn find_nearest_centroid(vector: &Vec<f32>, centroids: &Vec<Vec<f32>>) -> usize {
+        let mut calculator = L2DistanceCalculator::new();
+        centroids
+            .iter()
+            .enumerate()
+            .min_by(|(_, a), (_, b)| {
+                let dist_a = calculator.calculate(&vector, &a);
+                let dist_b = calculator.calculate(&vector, &b);
+                dist_a.partial_cmp(&dist_b).unwrap()
+            })
+            .map(|(index, _)| index)
+            .unwrap()
+    }
+
+    fn search(&self, query: &[f32], k: usize) -> Vec<(usize, f32)> {
```

Comment:
use `IdWithScore`

Reply:
This would require mmap support I believe? I'll need to read more about that part of the code still...

Reply:
`IdWithScore` is just a struct of `id` and `score`

Reply:
also, you'll need to rebase and added `SearchContext` as additional parameter